File: src/net/zomis/tttultimate/TTBase.java
package net.zomis.tttultimate;

import java.util.Collections;
import java.util.List;


public class TTBase implements Winnable, HasSub<TTBase> {
	// Container
	private final TTBase[][] subs;
	private final TTMNKParameters mnkParams;
	private final List<TTWinCondition> winConditions;
	
	// Winnable
	private final TTBase parent;
	private final int x;
	private final int y;
	private TTPlayer playedBy = TTPlayer.NONE;
	
	public TTBase(TTBase parent, TTMNKParameters parameters, TicFactory factory) {
		this(parent, 0, 0, parameters, factory);
	}
	public TTBase(TTBase parent, int x, int y, TTMNKParameters parameters, TicFactory factory) {
		this.parent = parent;
		this.mnkParams = parameters;
		this.x = x;
		this.y = y;
		
		this.subs = new TTBase[parameters.getWidth()][parameters.getHeight()];
		for (int xx = 0; xx < parameters.getWidth(); xx++) {
			for (int yy = 0; yy < parameters.getHeight(); yy++) {
				this.subs[xx][yy] = factory.construct(this, xx, yy);
			}
		}
		this.winConditions = Collections.unmodifiableList(TicUtils.setupWins(this));
	}
	
	public void determineWinner() {
		TTPlayer winner = TTPlayer.NONE;
		for (TTWinCondition cond : this.winConditions) {
			winner = winner.or(cond.determineWinnerNew());
		}
		this.playedBy = winner;
	}
	
	@Override
	public TTBase getSub(int x, int y) {
		if (!hasSubs() && x == 0 && y == 0)
			return this;
		if (x < 0 || y < 0)
			return null;
		if (x >= getSizeX() || y >= getSizeY())
			return null;
		return subs[x][y];
	}

	@Override
	public List<TTWinCondition> getWinConds() {
		return winConditions;
	}

	@Override
	public TTPlayer getWonBy() {
		return this.playedBy;
	}

	@Override
	public int getSizeX() {
		return this.mnkParams.getWidth();
	}

	@Override
	public int getSizeY() {
		return this.mnkParams.getHeight();
	}

	@Override
	public int getConsecutiveRequired() {
		return this.mnkParams.getConsecutiveRequired();
	}

	public TTMNKParameters getMNKParameters() {
		return this.mnkParams;
	}

	public TTBase getParent() {
		return parent;
	}
	
	public int getX() {
		return x;
	}
	public int getY() {
		return y;
	}
	
	public boolean isWon() {
		return playedBy != TTPlayer.NONE;
	}
	
	public void setPlayedBy(TTPlayer playedBy) {
		if (playedBy == null && this.hasSubs() && parent != null)
			new Exception().printStackTrace();
		this.playedBy = playedBy;
	}
	
	public boolean hasSubs() {
		return getSizeX() != 0 && getSizeY() != 0;
	}
	
	@Override
	public String toString() {
		return "{Pos " + x + ", " + y + "; Size " + getSizeX() + ", " + getSizeY() + "; Played by " + getWonBy() + ". Parent is " + parent + "}";
	}
	
	public void reset() {
		this.setPlayedBy(TTPlayer.NONE);
		for (int xx = 0; xx < getSizeX(); xx++) {
			for (int yy = 0; yy < getSizeY(); yy++) {
				this.getSub(xx, yy).reset();
			}
		}
	}
	public int getGlobalX() {
		if (parent == null)
			return 0;
		if (parent.getParent() == null)
			return x;
		return parent.getX() * parent.getParent().getSizeX() + this.x;
	}
	
	public int getGlobalY() {
		if (parent == null)
			return 0;
		if (parent.getParent() == null)
			return y;
		return parent.getY() * parent.getParent().getSizeY() + this.y;
	}
	
	public TTBase getSmallestTile(int x, int y) {
		int subX = x / getSizeX();
		int subY = y / getSizeY();
		TTBase board = getSub(subX, subY);
		if (board == null)
			throw new NullPointerException("No such smallest tile found: " + x + ", " + y);
		
		return board.getSub(x - subX*getSizeX(), y - subY*getSizeY());
	}
	
}

File: uttt-common/src/main/kotlin/net/zomis/tttultimate/TTBase.kt
package net.zomis.tttultimate

class TTBase(val parent: TTBase?, val x: Int, val y: Int,
     val mnkParameters: TTMNKParameters, factory: TicFactory?) : Winnable, HasSub<TTBase> {

    private val subs: Array<Array<TTBase>>

    override val winConds: List<TTWinCondition>

    override var wonBy: TTPlayer = TTPlayer.NONE

    override val sizeX: Int
        get() = this.mnkParameters.width

    override val sizeY: Int
        get() = this.mnkParameters.height

    override val consecutiveRequired: Int
        get() = this.mnkParameters.consecutiveRequired

    val isWon: Boolean
        get() = wonBy !== TTPlayer.NONE

    val globalX: Int
        get() {
            if (parent == null) {
                return 0
            }
            return if (parent.parent == null) x else parent.x * parent.parent.sizeX + this.x
        }

    val globalY: Int
        get() {
            if (parent == null) {
                return 0
            }
            return if (parent.parent == null) y else parent.y * parent.parent.sizeY + this.y
        }

    constructor(parent: TTBase?, parameters: TTMNKParameters, factory: TicFactory) : this(parent, 0, 0, parameters, factory)

    init {
        this.subs = Array(mnkParameters.width) { xx ->
            Array(mnkParameters.height) { yy ->
                factory!!.invoke(this, xx, yy)
            }
        }
        this.winConds = TicUtils.setupWins(this)
    }

    fun determineWinner() {
        var winner: TTPlayer = TTPlayer.NONE
        for (cond in this.winConds) {
            winner = winner.or(cond.determineWinnerNew())
        }
        if (winner == TTPlayer.NONE && subs().all { it.isWon }) {
            winner = TTPlayer.BLOCKED
        }
        this.wonBy = winner
    }

    fun subs(): List<TTBase> {
        if (!hasSubs()) {
            return emptyList()
        }
        return subs.flatMap { it.toList() }
    }

    override fun getSub(x: Int, y: Int): TTBase? {
        if (!hasSubs() && x == 0 && y == 0) {
            return this
        }
        if (x < 0 || y < 0) {
            return null
        }
        return if (x >= sizeX || y >= sizeY) null else subs[x][y]
    }

    fun setPlayedBy(playedBy: TTPlayer) {
        this.wonBy = playedBy
    }

    override fun hasSubs(): Boolean {
        return sizeX != 0 && sizeY != 0
    }

    override fun toString(): String {
        return "{Pos $x, $y; Size $sizeX, $sizeY; Played by $wonBy. Parent is $parent}"
    }

    fun reset() {
        this.setPlayedBy(TTPlayer.NONE)
        subs().forEach { it.reset() }
    }

    fun getSmallestTile(x: Int, y: Int): TTBase? {
        val topLeft = getSub(0, 0) ?: return null
        val grandParent = topLeft.hasSubs()

        if (!grandParent) {
            return this.getSub(x, y)
        }

        val subX = x / topLeft.sizeX
        val subY = y / topLeft.sizeY
        val board = getSub(subX, subY) ?: throw NullPointerException("No such smallest tile found: $x, $y")
        return board.getSub(x - subX * sizeX, y - subY * sizeY)
    }

}
